js/src/components/free-listings/configure-product-listings/checkErrors.test.js

```diff
@@ -0,0 +1,352 @@
+/**
+ * Internal dependencies
+ */
+import checkErrors from './checkErrors';
+
+function toRates( ...tuples ) {
+	return tuples.map( ( [ countryCode, rate ] ) => ( {
+		countryCode,
+		rate,
+	} ) );
+}
+
+function toTimes( ...tuples ) {
+	return tuples.map( ( [ countryCode, time ] ) => ( {
+		countryCode,
+		time,
+	} ) );
+}
+
+describe( 'checkErrors', () => {
+	describe( 'Shipping rates', () => {
+		let flatShipping;
+		let manualShipping;
+
+		beforeEach( () => {
+			flatShipping = { shipping_rate: 'flat' };
+			manualShipping = { shipping_rate: 'manual' };
+		} );
+
+		it( 'When the type of shipping rate is an invalid value, should not pass', () => {
+			// Not set yet
+			let errors = checkErrors( {}, [], [], [] );
+
+			expect( errors ).toHaveProperty( 'shipping_rate' );
+			expect( errors.shipping_rate ).toMatchSnapshot();
+
+			// Invalid value
+			errors = checkErrors( { shipping_rate: true }, [], [], [] );
+
+			expect( errors ).toHaveProperty( 'shipping_rate' );
+			expect( errors.shipping_rate ).toMatchSnapshot();
+
+			// Invalid value
+			errors = checkErrors( { shipping_rate: 'automatic' }, [], [], [] );
+
+			expect( errors ).toHaveProperty( 'shipping_rate' );
+			expect( errors.shipping_rate ).toMatchSnapshot();
+		} );
+
+		it( 'When the type of shipping rate is a valid value, should pass', () => {
+			// Selected flat
+			let errors = checkErrors( flatShipping, [], [], [] );
+
+			expect( errors ).not.toHaveProperty( 'shipping_rate' );
+
+			// Selected manual
+			errors = checkErrors( manualShipping, [], [], [] );
+
+			expect( errors ).not.toHaveProperty( 'shipping_rate' );
+		} );
+
+		describe( 'For flat type', () => {
+			function createFreeShipping( threshold ) {
+				return {
+					...flatShipping,
+					offers_free_shipping: true,
+					free_shipping_threshold: threshold,
+				};
+			}
+
+			it( `When there are any selected countries' shipping rates is not set, should not pass`, () => {
+				const rates = toRates( [ 'US', 10.5 ], [ 'FR', 12.8 ] );
+				const codes = [ 'US', 'JP', 'FR' ];
+
+				const errors = checkErrors( flatShipping, rates, [], codes );
+
+				expect( errors ).toHaveProperty( 'shipping_rate' );
+				expect( errors.shipping_rate ).toMatchSnapshot();
+			} );
+
+			it( `When all selected countries' shipping rates are set, should pass`, () => {
+				const rates = toRates( [ 'US', 10.5 ], [ 'FR', 12.8 ] );
+				const codes = [ 'US', 'FR' ];
+
+				const errors = checkErrors( flatShipping, rates, [], codes );
+
+				expect( errors ).not.toHaveProperty( 'shipping_rate' );
```

Comment:
:nail_care: 
I'm not sure how it affects tests atomicity:
With the way, I understand what "pass" means for `checkErrors` in general - there is no property with an error message.
Then we should either assert for `errors` not to have **any** property or change the description to specify there should be no rate error `shipping_rate`.

I could imagine :somebo: could expect that in some case `errors` could have another property here like `errors.missing_country`.

Reply:
Adjusted comment by https://github.com/woocommerce/google-listings-and-ads/pull/871/commits/5101eb382c404e929cb0efaab698b064f8af418e and added test case by https://github.com/woocommerce/google-listings-and-ads/pull/871/commits/2e4494a9fa0309998e4640aa48ea8dc91c618d9e
(:somebo: 😂 )